core/src/serve_repair.rs

```diff
@@ -506,6 +524,36 @@ impl ServeRepair {
         Ok(())
     }
 
+    fn repair_stake_stats(
+        &self,
+        bytes_requested: &HashMap<Pubkey, /*bytes*/ u64>,
+    ) -> Option<RepairStakeStats> {
+        let root_bank = self.bank_forks.read().unwrap().root_bank();
+        let epoch_schedule = root_bank.epoch_schedule();
+        let epoch = epoch_schedule.get_epoch(root_bank.slot());
+        root_bank.epoch_staked_nodes(epoch).map(|staked_nodes| {
+            let (staked, unstaked): (Vec<_>, Vec<_>) = bytes_requested
+                .iter()
+                .partition(|(id, _)| staked_nodes.contains_key(id));
+            let staked_peer_count = staked.len();
+            let unstaked_peer_count = unstaked.len();
+            let staked_response_bytes: u64 = staked.iter().map(|(_, x)| *x).sum();
+            let unstaked_response_bytes: u64 = unstaked.iter().map(|(_, x)| *x).sum();
+            let max_staked_peer_response_bytes: u64 =
+                staked.iter().map(|(_, x)| **x).max().unwrap_or_default();
+            let max_unstaked_peer_response_bytes: u64 =
+                unstaked.iter().map(|(_, x)| **x).max().unwrap_or_default();
```

Comment:
Buffering into vectors and repeated iterations here seems unnecessary.
Just loop over entries and update the metrics with a single iteration:
```rust
for (pubkey, bytes) in  bytes_requested {
   if staked_nodes.contains_key(pubkey) {
     staked_peer_count += 1;
     staked_response_bytes += bytes;
     max_staked_peer_response_bytes = ...;
   } else {
     // ...
   }
}

Reply:
done